grpc/controlproxy/controlproxy.go

```diff
@@ -0,0 +1,248 @@
+	req := depsRequest(taskID, mts, plugins)
+	reply, err := c.Client.SubscribeDeps(context.Background(), req)
+	if err != nil {
+		return []serror.SnapError{serror.New(err)}
+	}
+	serrs := common.ConvertSnapErrors(reply.Errors)
+	return serrs
+}
+
+func (c ControlProxy) UnsubscribeDeps(taskID string, mts []core.Metric, plugins []core.Plugin) []serror.SnapError {
+	req := depsRequest(taskID, mts, plugins)
+	reply, err := c.Client.UnsubscribeDeps(context.Background(), req)
+	if err != nil {
+		return []serror.SnapError{serror.New(err)}
+	}
+	serrs := common.ConvertSnapErrors(reply.Errors)
+	return serrs
+}
+
+func (c ControlProxy) MatchQueryToNamespaces(namespace []string) ([][]string, serror.SnapError) {
+	req := &rpc.ExpandWildcardsRequest{
+		Namespace: namespace,
+	}
+	reply, err := c.Client.MatchQueryToNamespaces(context.Background(), req)
+	if err != nil {
+		return nil, serror.New(err)
+	}
+	if reply.Error != nil {
+		return nil, common.ToSnapError(reply.Error)
+	}
+	nss := ToNSS(reply.NSS)
+	return nss, nil
+}
+
+///---------Util-------------------------------------------------------------------------
+// TODO(CDR): move the functions used by outside callers only to a util package and make functions used only
+// in this file non-exported.
+func getPluginType(t core.PluginType) int32 {
+	val := int32(-1)
+	switch t {
+	case core.CollectorPluginType:
+		val = 0
+	case core.ProcessorPluginType:
+		val = 1
+	case core.PublisherPluginType:
+		val = 2
+	}
+	return val
+}
+
+func depsRequest(taskID string, mts []core.Metric, plugins []core.Plugin) *rpc.SubscribeDepsRequest {
+	req := &rpc.SubscribeDepsRequest{
+		Metrics: common.NewMetrics(mts),
+		Plugins: common.CorePluginsToMsg(plugins),
+		TaskId:  taskID,
+	}
+	return req
+}
+
+func ConvertNSS(nss [][]string) []*rpc.ArrString {
+	res := make([]*rpc.ArrString, len(nss))
+	for i := range nss {
+		var tmp rpc.ArrString
+		tmp.S = nss[i]
+		res[i] = &tmp
+	}
+	return res
+}
+
+func ToNSS(arr []*rpc.ArrString) [][]string {
+	nss := make([][]string, len(arr))
+	for i, v := range arr {
+		nss[i] = v.S
+	}
+	return nss
+}
+
+func ErrorsToStrings(in []error) []string {
+	if len(in) == 0 {
+		return []string{}
+	}
+	erro := make([]string, len(in))
+	for i, e := range in {
+		erro[i] = e.Error()
+	}
+	return erro
+}
+
+func ReplyErrorsToErrors(errs []string) []error {
+	if len(errs) == 0 {
+		return []error{}
+	}
+	var erro []error
+	for _, e := range errs {
+		erro = append(erro, errors.New(e))
+	}
+	return erro
+}
+
+func GetPubProcReq(contentType string, content []byte, pluginName string, pluginVersion int, config map[string]ctypes.ConfigValue, taskID string) *rpc.PubProcMetricsRequest {
+	newConfig := common.ToConfigMap(config)
+	request := &rpc.PubProcMetricsRequest{
+		ContentType:   contentType,
+		Content:       content,
+		PluginName:    pluginName,
+		PluginVersion: int64(pluginVersion),
+		Config:        newConfig,
+		TaskId:        taskID,
+	}
+	return request
+}
```

Comment:
The shim looks great.  In a separate PR I think we should consider adding support for communicating in-process locally (not through gRPC or at least discuss this potential further since this shim sets us up for it.)